stan/math/opencl/kernels/tri_multiply.hpp

```diff
@@ -0,0 +1,155 @@
+      // matrix multiply where the multiplitcation is split
+      // in tiles.
+      // In cases where a matrix is lower triangular
+      // its not necessary to multiply the elements
+      // over the diagonal, therefore those tiles
+      // in the matrix multiply can be skipped.
+      // With upper triangular matrices we dont need
+      // to multiply the elements under the diagonal,
+      // so thosee tiles can be skipped.
+      // The following code determines the start and
+      // end tile based on triangularity of the input matrices
+      // If no matrices are triangular the starting tile
+      // is 0 and the end tile is num_tiles-1 which
+      // is the same as in \link kernels/multiply.hpp multiply()
+      const int end_tile_A = !is_A_lower * (num_tiles - 1)
+                             + is_A_lower * (i / THREAD_BLOCK_SIZE);
+      const int end_tile_B = !is_B_upper * (num_tiles - 1)
+                             + is_B_upper * (j / THREAD_BLOCK_SIZE);
+      const int start_tile_A = is_A_upper * (i / THREAD_BLOCK_SIZE);
+      const int start_tile_B = is_B_lower * (j / THREAD_BLOCK_SIZE);
+      int start_tile = 0;
+      int end_tile = num_tiles - 1;
+      if (start_tile_A > start_tile_B) {
+        start_tile = start_tile_A;
+      } else {
+        start_tile = start_tile_B;
+      }
+      if (end_tile_A > end_tile_B) {
+        end_tile = end_tile_B;
+      } else {
+        end_tile = end_tile_A;
+      }
+      for (int tile_ind = start_tile; tile_ind <= end_tile; tile_ind++) {
+        const int tiled_i = THREAD_BLOCK_SIZE * tile_ind + thread_block_row;
+        const int tiled_j = THREAD_BLOCK_SIZE * tile_ind + thread_block_col;
+        // each thread copies WORK_PER_THREAD values to the local
+        // memory
+        for (int w = 0; w < WORK_PER_THREAD; w++) {
+          const int tiled_i = THREAD_BLOCK_SIZE * tile_ind + thread_block_row;
+          const int tiled_j = THREAD_BLOCK_SIZE * tile_ind + thread_block_col;
+          // For the tiles on the diagonal we can ignore the values over
+          // the diagonal if the matrix is lower triangular or under
+          // the diagonal if the matrix is upper triangular
+          if ((is_A_lower && ((tiled_j + w * THREAD_BLOCK_SIZE_COL) <= i))
+              || (is_A_upper && ((tiled_j + w * THREAD_BLOCK_SIZE_COL) >= i))
+              || (!is_A_lower && !is_A_upper)) {
+            A_local[thread_block_col + w * THREAD_BLOCK_SIZE_COL]
+                   [thread_block_row]
+                = A[(tiled_j + w * THREAD_BLOCK_SIZE_COL) * M + i];
+          } else {
+            A_local[thread_block_col + w * THREAD_BLOCK_SIZE_COL]
+                   [thread_block_row]
+                = 0.0;
+          }
+          if ((is_B_lower && ((j + w * THREAD_BLOCK_SIZE_COL) <= tiled_i))
+              || (is_B_upper && ((j + w * THREAD_BLOCK_SIZE_COL) >= tiled_i))
+              || (!is_B_lower && !is_B_upper)) {
```

Comment:
Do we not have a triangular copy kernel lying around somewhere? I guess we can't easily call it, is that right? I wonder if we should be including more helper functions in that helpers.cl file...

Reply:
Again this was my instinct on performance that this check is faster than queuing another kernel. I will double check.

Reply:
Turns out my instinct were right, adding another copy does negatively effect performance. 

But based on your idea, I added the triangular template parameter to the `sub_block` and that way Its actually faster than the original version. So a big thanks for that!!

Reply:
hah, nice!